File: sdk-python/obsvr/action_context.py
"""Strict canonical action context without raw arguments or sensitive content."""

from __future__ import annotations

import hashlib
from typing import Any, Dict, List

from .aarm_outcome import AARM_OUTCOMES
from .tool_pinning import _canonical_json_for_hash

ACTION_CONTEXT_SCHEMA = "obsvr-action-context-v1"
_HASH_LENGTH = 64
_HEX = frozenset("0123456789abcdef")
_OUTCOMES = frozenset(AARM_OUTCOMES)
_MAX_SAFE_INTEGER = 9_007_199_254_740_991


class ActionContextValidationError(ValueError):
    """The input cannot produce one unambiguous canonical context."""


def _record(value: Any, field: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise ActionContextValidationError(f"{field} must be an object")
    return value


def _exact_keys(value: Dict[str, Any], allowed: set[str], field: str) -> None:
    unknown = sorted(set(value) - allowed)
    if unknown:
        raise ActionContextValidationError(
            f"{field} contains unsupported field: {unknown[0]}"
        )


def _nonblank(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ActionContextValidationError(f"{field} must be a nonblank string")
    return value


def _optional_nonblank(value: Dict[str, Any], key: str, field: str) -> str | None:
    if key not in value:
        return None
    return _nonblank(value[key], field)


def _hash(value: Any, field: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != _HASH_LENGTH
        or any(char not in _HEX for char in value)
    ):
        raise ActionContextValidationError(
            f"{field} must be exactly 64 lowercase hexadecimal characters"
        )
    return value


def _normalized_set(value: Any, field: str) -> List[str]:
    if not isinstance(value, list):
        raise ActionContextValidationError(f"{field} must be an array")
    strings = [_nonblank(entry, f"{field}[{index}]") for index, entry in enumerate(value)]
    return sorted(set(strings), key=lambda item: tuple(ord(char) for char in item))


def _prior_action(value: Any, index: int) -> Dict[str, Any]:
    field = f"prior_actions[{index}]"
    item = _record(value, field)
    _exact_keys(
        item,
        {
            "sequence",
            "kind",
            "name",
            "outcome",
            "receipt_hash",
            "data_classifications",
        },
        field,
    )
    sequence = item.get("sequence")
    if (
        isinstance(sequence, bool)
        or not isinstance(sequence, int)
        or sequence < 0
        or sequence > _MAX_SAFE_INTEGER
    ):
        raise ActionContextValidationError(
            f"{field}.sequence must be a nonnegative safe integer"
        )
    outcome = item.get("outcome")
    if not isinstance(outcome, str) or outcome not in _OUTCOMES:
        raise ActionContextValidationError(f"{field}.outcome is unsupported")
    return {
        "sequence": sequence,
        "kind": _nonblank(item.get("kind"), f"{field}.kind"),
        "name": _nonblank(item.get("name"), f"{field}.name"),
        "outcome": outcome,
        "receipt_hash": _hash(item.get("receipt_hash"), f"{field}.receipt_hash"),
        "data_classifications": _normalized_set(
            item.get("data_classifications"), f"{field}.data_classifications"
        ),
    }


def build_action_context(input_value: Any) -> Dict[str, Any]:
    """Build a strict canonical document from structured, non-content facts."""
    root = _record(input_value, "action context")
    _exact_keys(
        root,
        {
            "agent_id",
            "intent",
            "agent_role",
            "privilege_scope",
            "current_action",
            "run_id",
            "session_id",
            "thread_id",
            "prior_actions",
        },
        "action context",
    )
    current = _record(root.get("current_action"), "current_action")
    _exact_keys(
        current,
        {"kind", "name", "arguments_hash", "target", "data_classifications"},
        "current_action",
    )

    agent: Dict[str, Any] = {
        "agent_id": _nonblank(root.get("agent_id"), "agent_id"),
        "intent": _nonblank(root.get("intent"), "intent"),
    }
    role = _optional_nonblank(root, "agent_role", "agent_role")
    if role is not None:
        agent["role"] = role
    if "privilege_scope" in root:
        agent["privilege_scope"] = _normalized_set(
            root["privilege_scope"], "privilege_scope"
        )

    action: Dict[str, Any] = {
        "kind": _nonblank(current.get("kind"), "current_action.kind"),
        "name": _nonblank(current.get("name"), "current_action.name"),
        "arguments_hash": _hash(
            current.get("arguments_hash"), "current_action.arguments_hash"
        ),
        "data_classifications": _normalized_set(
            current.get("data_classifications"),
            "current_action.data_classifications",
        ),
    }
    target = _optional_nonblank(current, "target", "current_action.target")
    if target is not None:
        action["target"] = target

    prior_input = root.get("prior_actions")
    if not isinstance(prior_input, list):
        raise ActionContextValidationError("prior_actions must be an array")
    prior_actions = [_prior_action(item, index) for index, item in enumerate(prior_input)]
    for previous, current_action in zip(prior_actions, prior_actions[1:]):
        if current_action["sequence"] <= previous["sequence"]:
            raise ActionContextValidationError(
                "prior_actions sequences must be strictly increasing in input order"
            )

    doc: Dict[str, Any] = {
        "schema": ACTION_CONTEXT_SCHEMA,
        "agent": agent,
        "action": action,
        "run_id": _nonblank(root.get("run_id"), "run_id"),
        "prior_actions": prior_actions,
    }
    session_id = _optional_nonblank(root, "session_id", "session_id")
    if session_id is not None:
        doc["session_id"] = session_id
    thread_id = _optional_nonblank(root, "thread_id", "thread_id")
    if thread_id is not None:
        doc["thread_id"] = thread_id
    return doc


def canonicalize_action_context(input_value: Any) -> str:
    return _canonical_json_for_hash(build_action_context(input_value))


def action_context_hash(input_value: Any) -> str:
    return hashlib.sha256(
        canonicalize_action_context(input_value).encode("utf-8")
    ).hexdigest()

File: sdk-typescript/src/governance/action-context.ts
import { sha256Hex } from '../policy/decision-record.js';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import { AARM_OUTCOMES, type AarmOutcome } from './aarm-outcome.js';

export const ACTION_CONTEXT_SCHEMA = 'obsvr-action-context-v1' as const;

const HASH_RE = /^[0-9a-f]{64}$/;
const OUTCOMES = new Set<string>(AARM_OUTCOMES);

export interface PriorActionInput {
  sequence: number;
  kind: string;
  name: string;
  outcome: AarmOutcome;
  receipt_hash: string;
  data_classifications: string[];
}

export interface ActionContextInput {
  agent_id: string;
  active_intents: string[];
  agent_role?: string;
  privilege_scope?: string[];
  current_action: {
    kind: string;
    name: string;
    arguments_hash: string;
    target?: string;
    data_classifications: string[];
    requested_scopes: string[];
  };
  run_id: string;
  session_id?: string;
  thread_id?: string;
  prior_actions: PriorActionInput[];
}

export interface ActionContextDocument {
  schema: typeof ACTION_CONTEXT_SCHEMA;
  agent: {
    agent_id: string;
    active_intents: string[];
    role?: string;
    privilege_scope?: string[];
  };
  action: {
    kind: string;
    name: string;
    arguments_hash: string;
    target?: string;
    data_classifications: string[];
    requested_scopes: string[];
  };
  run_id: string;
  session_id?: string;
  thread_id?: string;
  prior_actions: PriorActionInput[];
}

export class ActionContextValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ActionContextValidationError';
  }
}

function record(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new ActionContextValidationError(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function exactKeys(
  value: Record<string, unknown>,
  allowed: readonly string[],
  field: string,
): void {
  const allowedSet = new Set(allowed);
  const unknown = Object.keys(value).filter((key) => !allowedSet.has(key));
  if (unknown.length > 0) {
    throw new ActionContextValidationError(
      `${field} contains unsupported field: ${unknown.sort()[0]}`,
    );
  }
}

function nonblank(value: unknown, field: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new ActionContextValidationError(`${field} must be a nonblank string`);
  }
  return value;
}

function optionalNonblank(value: Record<string, unknown>, key: string, field: string): string | undefined {
  if (!Object.prototype.hasOwnProperty.call(value, key)) return undefined;
  return nonblank(value[key], field);
}

function hash(value: unknown, field: string): string {
  if (typeof value !== 'string' || !HASH_RE.test(value)) {
    throw new ActionContextValidationError(
      `${field} must be exactly 64 lowercase hexadecimal characters`,
    );
  }
  return value;
}

function compareUnicodeScalars(left: string, right: string): number {
  const a = Array.from(left, (char) => char.codePointAt(0) as number);
  const b = Array.from(right, (char) => char.codePointAt(0) as number);
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

function normalizedSet(value: unknown, field: string): string[] {
  if (!Array.isArray(value)) {
    throw new ActionContextValidationError(`${field} must be an array`);
  }
  const values = value.map((entry, index) => nonblank(entry, `${field}[${index}]`));
  values.sort(compareUnicodeScalars);
  return values.filter((entry, index) => index === 0 || entry !== values[index - 1]);
}

function priorAction(value: unknown, index: number): PriorActionInput {
  const item = record(value, `prior_actions[${index}]`);
  exactKeys(
    item,
    ['sequence', 'kind', 'name', 'outcome', 'receipt_hash', 'data_classifications'],
    `prior_actions[${index}]`,
  );
  if (
    typeof item.sequence !== 'number'
    || !Number.isSafeInteger(item.sequence)
    || item.sequence < 0
  ) {
    throw new ActionContextValidationError(
      `prior_actions[${index}].sequence must be a nonnegative safe integer`,
    );
  }
  if (typeof item.outcome !== 'string' || !OUTCOMES.has(item.outcome)) {
    throw new ActionContextValidationError(
      `prior_actions[${index}].outcome is unsupported`,
    );
  }
  return {
    sequence: item.sequence,
    kind: nonblank(item.kind, `prior_actions[${index}].kind`),
    name: nonblank(item.name, `prior_actions[${index}].name`),
    outcome: item.outcome as AarmOutcome,
    receipt_hash: hash(item.receipt_hash, `prior_actions[${index}].receipt_hash`),
    data_classifications: normalizedSet(
      item.data_classifications,
      `prior_actions[${index}].data_classifications`,
    ),
  };
}

/** Build the canonical context without admitting raw arguments or content. */
export function buildActionContext(input: ActionContextInput): ActionContextDocument {
  const root = record(input, 'action context');
  exactKeys(
    root,
    [
      'agent_id',
      'active_intents',
      'agent_role',
      'privilege_scope',
      'current_action',
      'run_id',
      'session_id',
      'thread_id',
      'prior_actions',
    ],
    'action context',
  );

  const current = record(root.current_action, 'current_action');
  exactKeys(
    current,
    [
      'kind',
      'name',
      'arguments_hash',
      'target',
      'data_classifications',
      'requested_scopes',
    ],
    'current_action',
  );

  const agent: ActionContextDocument['agent'] = {
    agent_id: nonblank(root.agent_id, 'agent_id'),
    active_intents: normalizedSet(root.active_intents, 'active_intents'),
  };
  if (agent.active_intents.length === 0) {
    throw new ActionContextValidationError('active_intents must not be empty');
  }
  const role = optionalNonblank(root, 'agent_role', 'agent_role');
  if (role !== undefined) agent.role = role;
  if (Object.prototype.hasOwnProperty.call(root, 'privilege_scope')) {
    agent.privilege_scope = normalizedSet(root.privilege_scope, 'privilege_scope');
  }

  const action: ActionContextDocument['action'] = {
    kind: nonblank(current.kind, 'current_action.kind'),
    name: nonblank(current.name, 'current_action.name'),
    arguments_hash: hash(current.arguments_hash, 'current_action.arguments_hash'),
    data_classifications: normalizedSet(
      current.data_classifications,
      'current_action.data_classifications',
    ),
    requested_scopes: normalizedSet(
      current.requested_scopes,
      'current_action.requested_scopes',
    ),
  };
  const target = optionalNonblank(current, 'target', 'current_action.target');
  if (target !== undefined) action.target = target;

  if (!Array.isArray(root.prior_actions)) {
    throw new ActionContextValidationError('prior_actions must be an array');
  }
  const priorActions = root.prior_actions.map(priorAction);
  for (let i = 1; i < priorActions.length; i++) {
    if (priorActions[i].sequence <= priorActions[i - 1].sequence) {
      throw new ActionContextValidationError(
        'prior_actions sequences must be strictly increasing in input order',
      );
    }
  }

  const doc: ActionContextDocument = {
    schema: ACTION_CONTEXT_SCHEMA,
    agent,
    action,
    run_id: nonblank(root.run_id, 'run_id'),
    prior_actions: priorActions,
  };
  const sessionId = optionalNonblank(root, 'session_id', 'session_id');
  if (sessionId !== undefined) doc.session_id = sessionId;
  const threadId = optionalNonblank(root, 'thread_id', 'thread_id');
  if (threadId !== undefined) doc.thread_id = threadId;
  return doc;
}

export function canonicalizeActionContext(input: ActionContextInput): string {
  return canonicalJsonForHash(buildActionContext(input));
}

export function actionContextHash(input: ActionContextInput): string {
  return sha256Hex(canonicalizeActionContext(input));
}
